traits/trait_list_object.py

```diff
@@ -53,460 +88,755 @@ def __init__(self, index=0, removed=None, added=None):
+        the sort order, and the reverse flag can be set to request the result
+        in descending order.
+
+        Parameters
+        ----------
+        key : Callable
+            Custom function.
+        reverse : bool
+            If true, the resulting list will be sorted in descending order.
+
+        Notes
+        -----
+        Notification:
+            index : slice
+                The slice between 0 and len(self)
+            added : list
+                Will be [].
+            removed : list
+                Will be []
+
+        """
+
+        self[:] = sorted(self, key=key, reverse=reverse)
+
+        index = slice(0, len(self), None)
+        added = []
+        removed = []
+
+        self.notify(index, removed, added)
 
     def reverse(self):
-        removed = self[:]
-        if len(self) > 1:
-            list.reverse(self)
-            if self.name_items is not None:
-                self._send_trait_items_event(
-                    self.name_items, TraitListEvent(0, removed, self[:])
-                )
+        """ Reverse the order of the items in the list *IN PLACE*.
+
+        Notes
+        -----
+        Notification:
+            index : slice
+                The slice between 0 and len(self).
+            added : list
+                Will be []
+            removed : list.
+                Will be []
+
+        """
+        self[:] = self[::-1]
+
+        index = slice(0, len(self), None)
+        added = []
+        removed = []
+
+        self.notify(index, removed, added)
 
-    def pop(self, *args):
-        if not hasattr(self, "trait"):
-            return list.pop(self, *args)
-        if self.trait.minlen < len(self):
-            if len(args) > 0:
-                index = args[0]
+    def __iadd__(self, other):
+        """ Implements the self += value operator.
+
+        Parameters
+        ----------
+        other : Any
+            The item to add to the list.
+
+        Notes
+        -----
+        Notification:
+            index : slice
+                Slice ranging from current_length to new_length.
+            added : list
+                The  new values that were added.
+            removed : list
+                The empty list [].
+
+        """
+        self.extend(other)
+        return self
+
+    def __imul__(self, count):
+        """ Implements the self *= value operator.
+
+        Parameters
+        ----------
+        count : int
+            Number of times to duplicate the list.
+
+        Notes
+        -----
+        Notification:
+            index : slice
+                Slice ranging from current_length to new_length.
+            added : list
+                The  new values that were added.
+            removed : list
+                The empty list [].
+
+        """
+        if count > 1:
+            self.extend(self * (count - 1))
+        elif count == 0:
+            self[:] = []
+        return self
```

Comment:
> For TraitList (without minlen and maxlen), multiplying with a negative count should proceed as if clear() was called because that is what Python list does.

In case this came across as a suggestion to call `clear()` if count is negative, we need to handle non-int as well:
```
>>> x = [1, 2, 3]
>>> x * -10.2
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
TypeError: can't multiply sequence by non-int of type 'float'
```
One alternative (as is done in the old list) is to just call out to `super().__imul__(count)` for doing the actual modification.

Reply:
Reproducing the `TraitError` is going to be hard, a -ve length doesn't make sense, can we call this a bug and just fix it ?

Reply:
That also sounds reasonable.

Reply:
To confirm, adding this to `test_list.py`:
```

    def test_imul_negative(self):
        f = Foo()
        f.l = list("123")
        f.l *= -4
        self.assertEqual(f.l, [])
```

This test fails **on master**:
```
Traceback (most recent call last):
  File "/Users/kchoi/Work/ETS/traits/traits/tests/test_list.py", line 314, in test_imul_negative
    f.l *= -4
  File "/Users/kchoi/Work/ETS/traits/traits/trait_list_object.py", line 282, in __imul__
    self.len_error(original_len * count)
  File "/Users/kchoi/Work/ETS/traits/traits/trait_list_object.py", line 497, in len_error
    "s"[len == 1:],
traits.trait_errors.TraitError: The 'l' trait of a Foo instance must be a list of items which are a string, but you attempted to change its length to -12 elements.
```
but the test passes here.

Reply:
Negative count is handled. Resolving.